{"metadata":{"kernelspec":{"language":"python","display_name":"Python 3","name":"python3"},"language_info":{"pygments_lexer":"ipython3","nbconvert_exporter":"python","version":"3.6.4","file_extension":".py","codemirror_mode":{"name":"ipython","version":3},"name":"python","mimetype":"text/x-python"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"markdown","source":"# Essential Libraries","metadata":{}},{"cell_type":"code","source":"# Importing essential libraries for data manipulation and visualization\nimport numpy as np\nimport pandas as pd\nimport matplotlib.pyplot as plt\nimport seaborn as sns\nimport os\nimport time\n\n# Start time tracking for the notebook execution\nt0start = time.time()\n\n# List available data files\ndata_files = [os.path.join(dirname, filename) for dirname, _, filenames in os.walk('/kaggle/input') for filename in filenames]\ndata_files\n","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:16:52.511402Z","iopub.execute_input":"2023-09-24T03:16:52.511796Z","iopub.status.idle":"2023-09-24T03:16:54.959718Z","shell.execute_reply.started":"2023-09-24T03:16:52.511766Z","shell.execute_reply":"2023-09-24T03:16:54.958327Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Training Data","metadata":{}},{"cell_type":"code","source":"# Loading the training dataset\nfile_path = '/kaggle/input/open-problems-single-cell-perturbations/de_train.parquet'\ndf_de_train = pd.read_parquet(file_path)\ndf_de_train.shape, df_de_train.head()\n","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:16:54.962355Z","iopub.execute_input":"2023-09-24T03:16:54.963204Z","iopub.status.idle":"2023-09-24T03:16:57.592024Z","shell.execute_reply.started":"2023-09-24T03:16:54.963163Z","shell.execute_reply":"2023-09-24T03:16:57.591182Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# Previewing the initial columns of the training dataset\ndf_de_train.iloc[:, 5:].head(1)\n","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:16:57.593127Z","iopub.execute_input":"2023-09-24T03:16:57.593642Z","iopub.status.idle":"2023-09-24T03:16:57.654736Z","shell.execute_reply.started":"2023-09-24T03:16:57.593612Z","shell.execute_reply":"2023-09-24T03:16:57.653377Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Dimensional reductions (pca, umap,...), visualizations, clustering train target data","metadata":{}},{"cell_type":"code","source":"X = df_de_train.iloc[:,5:]\nprint(X.shape)","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:16:57.656210Z","iopub.execute_input":"2023-09-24T03:16:57.656595Z","iopub.status.idle":"2023-09-24T03:16:57.692059Z","shell.execute_reply.started":"2023-09-24T03:16:57.656553Z","shell.execute_reply":"2023-09-24T03:16:57.690920Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"%%time\nfrom sklearn.decomposition import PCA\n\nv1_color = df_de_train[  'cell_type']\nv2_color = df_de_train[  'sm_name'].copy()\nv3_color = df_de_train[  'sm_name'].copy()\nl = [t for t in df_de_train[  'sm_name'] if t.endswith('nib') ]\nm = v2_color.isin( l)\nv2_color[~m] = 'non -nib'\nv3_color[m] = '*nib'\n\nv4_color = df_de_train[  'control']#.copy()\n\nlist_top_drugs = ['MLN 2238', 'Resminostat', 'CEP-18770 (Delanzomib)', 'Oprozomib (ONX 0912)', 'Belinostat', 'Vorinostat', 'Ganetespib (STA-9090)', 'Scriptaid', 'Proscillaridin A;Proscillaridin-A', 'Alvocidib', 'IN1451']\nm = df_de_train[  'sm_name'].isin( list_top_drugs)\nv5_color = df_de_train[  'sm_name'].copy()\nv5_color[~m] = np.nan\n\n\n\nlist_cfg = [ ['cell type',v1_color], ['control' , v4_color ] , ['top compounds',v5_color] ]\n#     str_inf1 = ''\n    #X = np.clip(df.iloc[N0:N1,33:137].fillna(0),0, 1)\nstr_inf = 'PCA' \nreducer = PCA(n_components=100 )\nXr = reducer.fit_transform(X)\nfor i,j in [[0,1],[0,2],[1,2],[3,4],[5,6],[7,8]]:\n    plt.figure(figsize = (20,10)); ic=0\n    for str_inf1, v_for_color in list_cfg: # , ['*nib compounds ',v2_color], ['non *nib compounds',v3_color ] ]:\n        ic+=1; plt.subplot(1,len(list_cfg),ic)\n        sns.scatterplot(x= Xr[:,i], y = Xr[:,j], hue =  v_for_color ,s = 100) # df['reads'])\n        plt.xlabel(str_inf+str(i+1), fontsize = 20)\n        plt.ylabel(str_inf+str(j+1), fontsize = 20)\n        plt.title(str_inf1 + ' ', fontsize = 20 )\n\n    plt.show()","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:16:57.695285Z","iopub.execute_input":"2023-09-24T03:16:57.695889Z","iopub.status.idle":"2023-09-24T03:17:09.365421Z","shell.execute_reply.started":"2023-09-24T03:16:57.695857Z","shell.execute_reply":"2023-09-24T03:17:09.363940Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"d = df_de_train.iloc[:,:5]\nd['PCA1'] = Xr[:,0]\nd['PCA2'] = Xr[:,1]\nd['PCA3'] = Xr[:,2]\nlist_top_drugs = []\ndisplay( d.sort_values('PCA1', ascending = False ).head(8) )\nlist_top_drugs += d.sort_values('PCA1', ascending = False ).head(8)['sm_name'].to_list()\nprint(list_top_drugs)\ndisplay( d.sort_values('PCA2', ascending = False ).head(8) )\nlist_top_drugs += d.sort_values('PCA2', ascending = False ).head(8)['sm_name'].to_list()\ndisplay( d.sort_values('PCA3', ascending = False ).head(8) )\nlist_top_drugs += d.sort_values('PCA3', ascending = False ).head(8)['sm_name'].to_list()\nprint(list(set(list_top_drugs)))","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:17:09.367588Z","iopub.execute_input":"2023-09-24T03:17:09.368086Z","iopub.status.idle":"2023-09-24T03:17:09.431202Z","shell.execute_reply.started":"2023-09-24T03:17:09.368044Z","shell.execute_reply":"2023-09-24T03:17:09.430033Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Clustering  Cell Types","metadata":{}},{"cell_type":"code","source":"%%time\nN = df_de_train.shape[1]# 5000\nprint(N)\nX = df_de_train[ ['cell_type'] + list(df_de_train.columns[5:N]) ].groupby('cell_type').median()\nprint(X.shape)\ncm = np.corrcoef(X)\nprint(cm[:3,:2])\ncm = np.abs(cm)\nl = list(X.index)# [df_de_train['sm_name'].iat[i] +' '+ df_de_train['cell_type'].iat[i]  for i in range(len(df_de_train))] # .columns[5:N]\ncm = pd.DataFrame(cm, index =l , columns = l )\nprint(cm.shape)\nsns.clustermap(cm,  annot=True, fmt=\".2f\", cmap=\"coolwarm\" )\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:17:09.433292Z","iopub.execute_input":"2023-09-24T03:17:09.433752Z","iopub.status.idle":"2023-09-24T03:17:10.605278Z","shell.execute_reply.started":"2023-09-24T03:17:09.433710Z","shell.execute_reply":"2023-09-24T03:17:10.604103Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Clustering compounds","metadata":{}},{"cell_type":"code","source":"%%time\nN = df_de_train.shape[1]# 5000\nprint(N)\nX = df_de_train[ ['sm_name'] + list(df_de_train.columns[5:N]) ].groupby('sm_name').median()\nprint(X.shape)\ncm = np.corrcoef(X)\nprint(cm[:3,:2])\ncm = np.abs(cm)\nl = list(X.index)# [df_de_train['sm_name'].iat[i] +' '+ df_de_train['cell_type'].iat[i]  for i in range(len(df_de_train))] # .columns[5:N]\nl = [t[:20] for t in l] # cut long names\ncm = pd.DataFrame(cm, index =l , columns = l )\nprint(cm.shape)\nclustergrid = sns.clustermap(cm,cmap=\"coolwarm\" )# ,  annot=True, fmt=\".2f\", \nplt.show()\nreordered_columns = clustergrid.dendrogram_col.reordered_ind\nreordered_rows = clustergrid.dendrogram_row.reordered_ind\nprint(len(reordered_rows), len(reordered_columns) )\nprint( list(cm.index[reordered_rows]) )\n# print( list(X.columns[reordered_columns]) )\n\nsns.clustermap(cm,  annot=True, fmt=\".2f\", cmap=\"coolwarm\" )\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:17:10.607015Z","iopub.execute_input":"2023-09-24T03:17:10.607422Z","iopub.status.idle":"2023-09-24T03:18:00.104221Z","shell.execute_reply.started":"2023-09-24T03:17:10.607379Z","shell.execute_reply":"2023-09-24T03:18:00.102966Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Clustering samples (i.e. pairs cell + compound)","metadata":{}},{"cell_type":"code","source":"%%time\nN = df_de_train.shape[1]# 5000\nX = df_de_train.iloc[:,5:N]\nprint(X.shape)\ncm = np.corrcoef(X)\nprint(cm[:3,:2])\ncm = np.abs(cm)\nl = [df_de_train['sm_name'].iat[i] +' '+ df_de_train['cell_type'].iat[i]  for i in range(len(df_de_train))] # .columns[5:N]\ncm = pd.DataFrame(cm, index =l , columns = l )\nprint(cm.shape)\nclustergrid = sns.clustermap(cm)\nplt.show()\nreordered_columns = clustergrid.dendrogram_col.reordered_ind\nreordered_rows = clustergrid.dendrogram_row.reordered_ind\nprint(len(reordered_rows), len(reordered_columns) )\nprint( list(cm.index[reordered_rows]) )\n# print( list(cm.columns[reordered_columns]) )","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:18:00.105970Z","iopub.execute_input":"2023-09-24T03:18:00.106412Z","iopub.status.idle":"2023-09-24T03:18:02.400116Z","shell.execute_reply.started":"2023-09-24T03:18:00.106376Z","shell.execute_reply":"2023-09-24T03:18:02.399059Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Look in genes space","metadata":{}},{"cell_type":"code","source":"%%time\nfrom sklearn.decomposition import PCA\n\nX = df_de_train.iloc[:,5:].T\nprint(X.shape)\n\nv1_color = pd.Series(range(len(X)), name = 'index') # df_de_train[  'cell_type']\n\n\n\nlist_cfg = [ ['Genes',v1_color]]# , ['control' , v4_color ] , ['top compounds',v5_color] ]\n#     str_inf1 = ''\n    #X = np.clip(df.iloc[N0:N1,33:137].fillna(0),0, 1)\nstr_inf = 'PCA' \nreducer = PCA(n_components=10 )\nXr = reducer.fit_transform(X)\nfor i,j in [[0,1],[0,2],[1,2],[3,4],[5,6],[7,8]]:\n    plt.figure(figsize = (20,10)); ic=0\n    for str_inf1, v_for_color in list_cfg: # , ['*nib compounds ',v2_color], ['non *nib compounds',v3_color ] ]:\n        ic+=1; plt.subplot(1,len(list_cfg),ic)\n        sns.scatterplot(x= Xr[:,i], y = Xr[:,j], hue =  v_for_color ,s = 100) # df['reads'])\n        plt.xlabel(str_inf+str(i+1), fontsize = 20)\n        plt.ylabel(str_inf+str(j+1), fontsize = 20)\n        plt.title(str_inf1 + ' ', fontsize = 20 )\n\n    plt.show()","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:18:02.402177Z","iopub.execute_input":"2023-09-24T03:18:02.402629Z","iopub.status.idle":"2023-09-24T03:18:14.538063Z","shell.execute_reply.started":"2023-09-24T03:18:02.402592Z","shell.execute_reply":"2023-09-24T03:18:14.536779Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"%%time\nfrom sklearn.decomposition import PCA\nimport umap \n\nX = df_de_train.iloc[:,5:].T\nprint(X.shape)\n\nv1_color = pd.Series(range(len(X)), name = 'index') # df_de_train[  'cell_type']\n\n\n\nlist_cfg = [ ['Genes',v1_color]]# , ['control' , v4_color ] , ['top compounds',v5_color] ]\n# str_inf = 'PCA' \n# reducer = PCA(n_components=10 )\nstr_inf = 'UMAP' \nreducer = umap.UMAP()# (n_components=10 )\n\nXr = reducer.fit_transform(X)\nfor i,j in [[0,1] ]:# ,[0,2],[1,2],[3,4],[5,6],[7,8]]:\n    plt.figure(figsize = (20,10)); ic=0\n    for str_inf1, v_for_color in list_cfg: # , ['*nib compounds ',v2_color], ['non *nib compounds',v3_color ] ]:\n        ic+=1; plt.subplot(1,len(list_cfg),ic)\n        sns.scatterplot(x= Xr[:,i], y = Xr[:,j], hue =  v_for_color ,s = 100) # df['reads'])\n        plt.xlabel(str_inf+str(i+1), fontsize = 20)\n        plt.ylabel(str_inf+str(j+1), fontsize = 20)\n        plt.title(str_inf1 + ' ', fontsize = 20 )\n\n    plt.show()","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:18:14.540099Z","iopub.execute_input":"2023-09-24T03:18:14.540525Z","iopub.status.idle":"2023-09-24T03:19:29.766554Z","shell.execute_reply.started":"2023-09-24T03:18:14.540490Z","shell.execute_reply":"2023-09-24T03:19:29.765317Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Genes clustering","metadata":{}},{"cell_type":"code","source":"%%time\nN = 1000#  18211 #  15_000# 10000 #\nX = df_de_train.iloc[:,5:N].T\nprint(X.shape)\ncm = np.corrcoef(X)\nprint(cm[:3,:2])\ncm = np.abs(cm)\ncm = pd.DataFrame(cm, index = df_de_train.columns[5:N], columns = df_de_train.columns[5:N] )\nprint(cm.shape)\nsns.clustermap(cm)\nplt.show()","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:19:29.768360Z","iopub.execute_input":"2023-09-24T03:19:29.769426Z","iopub.status.idle":"2023-09-24T03:19:32.942034Z","shell.execute_reply.started":"2023-09-24T03:19:29.769384Z","shell.execute_reply":"2023-09-24T03:19:32.940663Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"G1S_genes_Tirosh = ['MCM5', 'PCNA', 'TYMS', 'FEN1', 'MCM2', 'MCM4', 'RRM1', 'UNG', 'GINS2', 'MCM6', 'CDCA7', 'DTL', 'PRIM1', 'UHRF1', 'MLF1IP', 'HELLS', 'RFC2', 'RPA2', 'NASP', 'RAD51AP1', 'GMNN', 'WDR76', 'SLBP', 'CCNE2', 'UBR7', 'POLD3', 'MSH2', 'ATAD2', 'RAD51', 'RRM2', 'CDC45', 'CDC6', 'EXO1', 'TIPIN', 'DSCC1', 'BLM', 'CASP8AP2', 'USP1', 'CLSPN', 'POLA1', 'CHAF1B', 'BRIP1', 'E2F8']\nG2M_genes_Tirosh = ['HMGB2', 'CDK1', 'NUSAP1', 'UBE2C', 'BIRC5', 'TPX2', 'TOP2A', 'NDC80', 'CKS2', 'NUF2', 'CKS1B', 'MKI67', 'TMPO', 'CENPF', 'TACC3', 'FAM64A', 'SMC4', 'CCNB2', 'CKAP2L', 'CKAP2', 'AURKB', 'BUB1', 'KIF11', 'ANP32E', 'TUBB4B', 'GTSE1', 'KIF20B', 'HJURP', 'CDCA3', 'HN1', 'CDC20', 'TTK', 'CDC25C', 'KIF2C', 'RANGAP1', 'NCAPD2', 'DLGAP5', 'CDCA2', 'CDCA8', 'ECT2', 'KIF23', 'HMMR', 'AURKA', 'PSRC1', 'ANLN', 'LBR', 'CKAP5', 'CENPE', 'CTCF', 'NEK2', 'G2E3', 'GAS2L3', 'CBX5', 'CENPA']\ngenes_Tirosh = G1S_genes_Tirosh + G2M_genes_Tirosh\n\n# Subset of Tirosh genes to capture \"fast\" cell cycle pattern = see https://arxiv.org/abs/2208.05229\nlist_genes_fastCCsign = ['CDK1', 'UBE2C', 'TOP2A', 'TMPO', 'HJURP', 'RRM1', 'RAD51AP1', 'RRM2', 'CDC45', 'BLM', 'BRIP1', 'E2F8', 'HIST2H2AC']\n\nG1S_genes_Freeman = ['ADAMTS1', 'ASF1B', 'ATAD2', 'BARD1', 'BLM', 'BRCA1', 'BRIP1', 'C17orf75', 'C9orf40', 'CACYBP', 'CASP8AP2', 'CCDC15', 'CCNE1', 'CCNE2', 'CCP110', 'CDC25A', 'CDC45', 'CDC6', 'CDC7', 'CDK2', 'CDT1', 'CENPJ', 'CENPQ', 'CENPU', 'CEP57', 'CHAF1A', 'CHAF1B', 'CHEK1', 'CLSPN', 'CREBZF', 'CRYL1', 'CSE1L', 'DCLRE1B', 'DCTPP1', 'DEK', 'DERA', 'DHFR', 'DNA2', 'DNAJC9', 'DNMT1', 'DONSON', 'DSCC1', 'DSN1', 'DTL', 'E2F8', 'EED', 'EFCAB11', 'ENDOD1', 'ETAA1', 'EXO1', 'EYA2', 'EZH2', 'FAM111A', 'FANCE', 'FANCG', 'FANCI', 'FANCL', 'FBXO5', 'FEN1', 'GGH', 'GINS1', 'GINS2', 'GINS3', 'GLMN', 'GMNN', 'GMPS', 'GPD2', 'HADH', 'HELLS', 'HSF2', 'ITGB3BP', 'KIAA0101', 'KNTC1', 'LIG1', 'MCM10', 'MCM2', 'MCM3', 'MCM4', 'MCM5', 'MCM6', 'MCM7', 'MCMBP', 'METTL9', 'MMD', 'MNS1', 'MPP1', 'MRE11A', 'MSH2', 'MSH6', 'MYO19', 'NASP', 'NPAT', 'NSMCE4A', 'ORC1', 'OSGEPL1', 'PAK1', 'PAQR4', 'PARP2', 'PASK', 'PAXIP1', 'PBX3', 'PCNA', 'PKMYT1', 'PMS1', 'POLA1', 'POLA2', 'POLD3', 'POLE2', 'PRIM1', 'PRPS2', 'PSMC3IP', 'RAB23', 'RAD51', 'RAD51AP1', 'RAD54L', 'RBBP8', 'RBL1', 'RDX', 'RFC2', 'RFC3', 'RFC4', 'RMI1', 'RNASEH2A', 'RPA1', 'RRM1', 'RRM2', 'SLBP', 'SLC25A40', 'SMC2', 'SMC3', 'SSX2IP', 'SUPT16H', 'TEX30', 'TFDP1', 'THAP10', 'THEM6', 'TIMELESS', 'TIPIN', 'TMEM106C', 'TMEM38B', 'TRIM45', 'TRIP13', 'TSPYL4', 'TTI1', 'TUBGCP5', 'TYMS', 'UBR7', 'UNG', 'USP1', 'WDHD1', 'WDR76', 'WRB', 'YEATS4', 'ZBTB14', 'ZWINT']\nG2M_genes_Freeman = ['ADGRE5', 'ARHGAP11A', 'ARHGDIB', 'ARL6IP1', 'ASPM', 'AURKA', 'AURKB', 'BIRC5', 'BORA', 'BRD8', 'BUB1', 'BUB1B', 'BUB3', 'CCNA2', 'CCNB1', 'CCNB2', 'CCNF', 'CDC20', 'CDC25B', 'CDC25C', 'CDC27', 'CDCA3', 'CDCA8', 'CDK1', 'CDKN1B', 'CDKN3', 'CENPE', 'CENPF', 'CENPI', 'CENPN', 'CEP55', 'CEP70', 'CEP85', 'CKAP2', 'CKAP5', 'CKS1B', 'CKS2', 'CTCF', 'DBF4', 'DBF4B', 'DCAF7', 'DEPDC1', 'DLGAP5', 'ECT2', 'ERCC6L', 'ESPL1', 'FAM64A', 'FOXM1', 'FZD2', 'FZD7', 'FZR1', 'GPSM2', 'GTF2E1', 'GTSE1', 'H2AFX', 'HJURP', 'HMGB2', 'HMGB3', 'HMMR', 'HN1', 'INCENP', 'JADE2', 'KIF11', 'KIF14', 'KIF15', 'KIF18A', 'KIF18B', 'KIF20A', 'KIF20B', 'KIF22', 'KIF23', 'KIF2C', 'KIF4A', 'KIF5B', 'KIFC1', 'KPNA2', 'LBR', 'LMNB2', 'MAD2L1', 'MELK', 'MET', 'METTL4', 'MIS18BP1', 'MKI67', 'MPHOSPH9', 'MTMR6', 'NCAPD2', 'NCAPG', 'NCAPG2', 'NCAPH', 'NDC1', 'NDC80', 'NDE1', 'NEIL3', 'NEK2', 'NRF1', 'NUSAP1', 'OIP5', 'PAFAH2', 'PARPBP', 'PBK', 'PLEKHG3', 'PLK1', 'PLK4', 'PRC1', 'PRR11', 'PSRC1', 'PTTG1', 'PTTG3P', 'RACGAP1', 'RAD21', 'RASSF1', 'REEP4', 'SAP30', 'SHCBP1', 'SKA1', 'SLCO1B3', 'SOGA1', 'SPA17', 'SPAG5', 'SPC25', 'SPDL1', 'STIL', 'STK17B', 'TACC3', 'TAF5', 'TBC1D2', 'TBC1D31', 'TMPO', 'TOP2A', 'TPX2', 'TROAP', 'TTF2', 'TTK', 'TUBB4B', 'TUBD1', 'UBE2C', 'UBE2S', 'VANGL1', 'WEE1', 'WHSC1', 'XPO1', 'ZMYM1']","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:19:32.943762Z","iopub.execute_input":"2023-09-24T03:19:32.944097Z","iopub.status.idle":"2023-09-24T03:19:32.966803Z","shell.execute_reply.started":"2023-09-24T03:19:32.944068Z","shell.execute_reply":"2023-09-24T03:19:32.965608Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"%%time\nN = 1000#  18211 #  15_000# 10000 #\n\nfor l,str_inf in [ [G1S_genes_Tirosh, 'G1S Tirosh'], [G2M_genes_Tirosh, 'G2M Tirosh'],  [G1S_genes_Tirosh + G2M_genes_Tirosh, 'All Tirosh'],\n                  [list_genes_fastCCsign, 'FastCC Signature'],\n                  [ G1S_genes_Freeman, 'G1S Freeman' ],  [ G2M_genes_Freeman, 'G2M Freeman' ], [ G1S_genes_Freeman + G2M_genes_Freeman, 'All Freeman' ],   ]: \n    ll = set(l) & set(df_de_train.columns) \n    ll = list(ll)\n    print(len(ll), str_inf )\n    X = df_de_train[ll].T # .iloc[:,5:N].T\n    print(X.shape)\n    cm = np.corrcoef(X)\n    print(cm[:3,:2])\n    cm = np.abs(cm)\n    cm = pd.DataFrame(cm, index = ll, columns = ll )\n    print(cm.shape)\n    clustergrid = sns.clustermap(cm)\n    plt.title(str_inf, fontsize = 20 )\n    plt.show()\n    reordered_columns = clustergrid.dendrogram_col.reordered_ind\n    reordered_rows = clustergrid.dendrogram_row.reordered_ind\n    print(len(reordered_rows), len(reordered_columns) )\n    print( list(cm.index[reordered_rows]) )\n#     print( list(cm.columns[reordered_columns]) )","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:19:32.973065Z","iopub.execute_input":"2023-09-24T03:19:32.973628Z","iopub.status.idle":"2023-09-24T03:19:41.026496Z","shell.execute_reply.started":"2023-09-24T03:19:32.973591Z","shell.execute_reply":"2023-09-24T03:19:41.025484Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"df_de_train","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:19:41.028213Z","iopub.execute_input":"2023-09-24T03:19:41.028652Z","iopub.status.idle":"2023-09-24T03:19:41.069962Z","shell.execute_reply.started":"2023-09-24T03:19:41.028615Z","shell.execute_reply":"2023-09-24T03:19:41.068594Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"%%time\n# for l,str_inf in [ [G1S_genes_Tirosh, 'G1S Tirosh'], [G2M_genes_Tirosh, 'G2M Tirosh'],  [G1S_genes_Tirosh + G2M_genes_Tirosh, 'All Tirosh'],\n#                   [list_genes_fastCCsign, 'FastCC Signature'],\n#                   [ G1S_genes_Freeman, 'G1S Freeman' ],  [ G2M_genes_Freeman, 'G2M Freeman' ], [ G1S_genes_Freeman + G2M_genes_Freeman, 'All Freeman' ],   ]: \nfor l,str_inf in [  [G1S_genes_Tirosh + G2M_genes_Tirosh, 'All Tirosh']   ]: \n    ll = set(l) & set(df_de_train.columns) \n    ll = list(ll)\n    print(len(ll), str_inf )\n    X = df_de_train[ ['cell_type'] + list(df_de_train.columns[5:]) ].groupby('cell_type').median()\n    X = X[ll]\n    print(X.shape)\n    clustergrid = sns.clustermap(X)# ,  annot=True, fmt=\".2f\", cmap=\"coolwarm\" )\n    plt.title(str_inf, fontsize = 20 )\n    plt.show()\n    reordered_columns = clustergrid.dendrogram_col.reordered_ind\n    reordered_rows = clustergrid.dendrogram_row.reordered_ind\n    print(len(reordered_rows), len(reordered_columns) )\n    print( list(X.index[reordered_rows]) )\n    print( list(X.columns[reordered_columns]) )\n    \ncol = 'sm_name'    \nfor l,str_inf in [  [G1S_genes_Tirosh + G2M_genes_Tirosh, 'All Tirosh']   ]: \n    ll = set(l) & set(df_de_train.columns) \n    ll = list(ll)\n    print(len(ll), str_inf )\n    X = df_de_train[ [col] + list(df_de_train.columns[5:]) ].groupby(col).median()\n    X = X[ll]\n    print(X.shape)\n    X.index = [t[:20] for t in X.index] # cut too long names\n    clustergrid = sns.clustermap(X)# ,  annot=True, fmt=\".2f\", cmap=\"coolwarm\" )\n    plt.title(str_inf, fontsize = 20 )\n    plt.show()    \n    reordered_columns = clustergrid.dendrogram_col.reordered_ind\n    reordered_rows = clustergrid.dendrogram_row.reordered_ind\n    print(len(reordered_rows), len(reordered_columns) )\n    print( list(X.index[reordered_rows]) )\n    print( list(X.columns[reordered_columns]) )","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:19:41.071877Z","iopub.execute_input":"2023-09-24T03:19:41.072234Z","iopub.status.idle":"2023-09-24T03:19:43.600757Z","shell.execute_reply.started":"2023-09-24T03:19:41.072203Z","shell.execute_reply":"2023-09-24T03:19:43.599862Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Look on compounds ( count = 146  )\n\n15 compounds - 6 times data - only in train","metadata":{}},{"cell_type":"code","source":"d = df_de_train[['sm_name','sm_lincs_id','SMILES']].drop_duplicates()\nprint(d.shape)\nd.to_csv('compounds.csv')\ndisplay( d.head(10) )\n\nprint(list(df_de_train['sm_name'].unique() ) )","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:19:43.602288Z","iopub.execute_input":"2023-09-24T03:19:43.602816Z","iopub.status.idle":"2023-09-24T03:19:43.626483Z","shell.execute_reply.started":"2023-09-24T03:19:43.602785Z","shell.execute_reply":"2023-09-24T03:19:43.625119Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"l = [len(s) for s in df_de_train['SMILES']]\nnp.sort(list(set(l)) )","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:19:43.628409Z","iopub.execute_input":"2023-09-24T03:19:43.628849Z","iopub.status.idle":"2023-09-24T03:19:43.637607Z","shell.execute_reply.started":"2023-09-24T03:19:43.628808Z","shell.execute_reply":"2023-09-24T03:19:43.636334Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"display( df_de_train['sm_name'].value_counts().head(20) )\ndisplay( df_de_train['sm_name'].value_counts().tail(10) )\ndf_de_train['sm_name'].value_counts().value_counts()","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:19:43.639685Z","iopub.execute_input":"2023-09-24T03:19:43.640126Z","iopub.status.idle":"2023-09-24T03:19:43.665832Z","shell.execute_reply.started":"2023-09-24T03:19:43.640086Z","shell.execute_reply":"2023-09-24T03:19:43.665014Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Aggregations by compounds, cell_types \n\nIt is used for prediction in early versions of the notebook","metadata":{}},{"cell_type":"code","source":"%%time\ntrain_aggregate_mean_or_median_or_whatever = df_de_train.iloc[:,5:].quantile(0.7)# median()\ntrain_aggregate_mean_or_median_or_whatever","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:19:43.666910Z","iopub.execute_input":"2023-09-24T03:19:43.667247Z","iopub.status.idle":"2023-09-24T03:19:44.119277Z","shell.execute_reply.started":"2023-09-24T03:19:43.667218Z","shell.execute_reply":"2023-09-24T03:19:44.118458Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"%%time\nd = train_aggregate_mean_or_median_or_whatever\nplt.figure(figsize = (20,4) )\nplt.plot(d.values)\nplt.show()\nplt.figure(figsize = (10,4) )\nplt.hist(d.values, bins = 100)\nplt.show()\n\ndisplay( d.describe() )","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:19:44.120328Z","iopub.execute_input":"2023-09-24T03:19:44.120650Z","iopub.status.idle":"2023-09-24T03:19:44.951217Z","shell.execute_reply.started":"2023-09-24T03:19:44.120622Z","shell.execute_reply":"2023-09-24T03:19:44.949953Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Submission data","metadata":{}},{"cell_type":"code","source":"%%time\nfn = '/kaggle/input/open-problems-single-cell-perturbations/id_map.csv'\ndf_id_map = pd.read_csv(fn)\nprint(df_id_map.shape)\ndisplay(df_id_map)\nfn = '/kaggle/input/open-problems-single-cell-perturbations/sample_submission.csv'\ndf = pd.read_csv(fn, index_col = 0)\nprint(df.shape)\ndf","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:19:44.952731Z","iopub.execute_input":"2023-09-24T03:19:44.953278Z","iopub.status.idle":"2023-09-24T03:19:49.806422Z","shell.execute_reply.started":"2023-09-24T03:19:44.953234Z","shell.execute_reply":"2023-09-24T03:19:49.805178Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Prepare submission by aggregated-target","metadata":{}},{"cell_type":"markdown","source":"## Key params","metadata":{}},{"cell_type":"code","source":"# predict_method = 'train_aggregation_by_compounds'\n# predict_method = 'train_aggregation_by_compounds_with_denoising_pca'\n# predict_method = 'train_aggregation_by_compounds_with_denoising_ICA'\npredict_method = 'train_aggregation_by_compounds_with_denoising_TSVD'\nquantile = 0.54\nn_components = 35","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:19:49.807950Z","iopub.execute_input":"2023-09-24T03:19:49.808398Z","iopub.status.idle":"2023-09-24T03:19:49.814208Z","shell.execute_reply.started":"2023-09-24T03:19:49.808369Z","shell.execute_reply":"2023-09-24T03:19:49.813429Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Prepare predictions by direct aggregation of targets by compounds","metadata":{}},{"cell_type":"code","source":"%%time\ntrain_aggr_direct = df_de_train[ ['sm_name'] + list(df_de_train.columns[5:])  ].groupby('sm_name' ).quantile(quantile)# median()\ntrain_aggr_direct","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:19:49.815855Z","iopub.execute_input":"2023-09-24T03:19:49.816291Z","iopub.status.idle":"2023-09-24T03:19:51.706753Z","shell.execute_reply.started":"2023-09-24T03:19:49.816233Z","shell.execute_reply":"2023-09-24T03:19:51.705127Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"## Prepapre predictions by aggregation of PCA-reduced targets by compounds","metadata":{}},{"cell_type":"code","source":"%%time \nfrom sklearn.decomposition import PCA\nfrom sklearn.decomposition import FastICA\nfrom sklearn.decomposition import TruncatedSVD\n\nY = df_de_train.iloc[:,5:]\nprint(X.shape)\nif '_pca' in predict_method:\n    str_inf_target_dimred = 'PCA' \n    reducer = PCA(n_components=n_components )\nelif '_ICA' in predict_method:\n    str_inf_target_dimred = 'ICA' \n#     reducer = PCA(n_components=n_components )\n    reducer = FastICA(n_components=n_components, random_state=0, whiten='unit-variance')\nelif '_TSVD' in predict_method:\n    str_inf_target_dimred = 'TSVD' \n#     reducer = PCA(n_components=n_components )\n#     reducer = FastICA(n_components=n_components, random_state=0, whiten='unit-variance')\n    reducer = TruncatedSVD(n_components=n_components, n_iter=7, random_state=42)\nelse:\n    str_inf_target_dimred = ''\n    \nprint(str_inf_target_dimred , reducer)\nYr = reducer.fit_transform(Y)\nYr_inv_trans = reducer.inverse_transform(Yr)\ndf_red_inv_trans = pd.DataFrame(Yr_inv_trans, columns = df_de_train.columns[5:])\ndf_red_inv_trans['sm_name'] = df_de_train['sm_name']\n\ntrain_aggr_denoised = df_red_inv_trans.groupby('sm_name' ).quantile(quantile)# median()\ntrain_aggr_denoised","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:19:51.708520Z","iopub.execute_input":"2023-09-24T03:19:51.708897Z","iopub.status.idle":"2023-09-24T03:19:56.178247Z","shell.execute_reply.started":"2023-09-24T03:19:51.708867Z","shell.execute_reply":"2023-09-24T03:19:56.177010Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"%%time\n\nif predict_method == 'train_aggregation_by_compounds':\n    df = df_id_map.merge(train_aggr, how = 'left', on = 'sm_name')\n    df = df.set_index('id').iloc[:,2:]\nelif predict_method.startswith('train_aggregation_by_compounds_with_denoising_'):\n    df = df_id_map.merge(train_aggr_denoised, how = 'left', on = 'sm_name')\n    df = df.set_index('id').iloc[:,2:]\nelse:\n    # consant for each target submission:\n    for i,col in enumerate( df.columns ):\n        df[col] = train_aggregate_mean_or_median_or_whatever[col]\n        if (i%1000) == 0: print(i,col)\n    \ndf","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:19:56.179828Z","iopub.execute_input":"2023-09-24T03:19:56.180311Z","iopub.status.idle":"2023-09-24T03:19:56.294861Z","shell.execute_reply.started":"2023-09-24T03:19:56.180282Z","shell.execute_reply":"2023-09-24T03:19:56.293775Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"%%time\ndf","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:19:56.296322Z","iopub.execute_input":"2023-09-24T03:19:56.296646Z","iopub.status.idle":"2023-09-24T03:19:56.336228Z","shell.execute_reply.started":"2023-09-24T03:19:56.296620Z","shell.execute_reply":"2023-09-24T03:19:56.335411Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"%%time\ndf.to_csv('submission.csv')","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:19:56.337763Z","iopub.execute_input":"2023-09-24T03:19:56.338093Z","iopub.status.idle":"2023-09-24T03:20:09.745345Z","shell.execute_reply.started":"2023-09-24T03:19:56.338062Z","shell.execute_reply":"2023-09-24T03:20:09.744108Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Towards modeling","metadata":{}},{"cell_type":"markdown","source":"## Key params","metadata":{}},{"cell_type":"code","source":"n_components_for_cell_type_encoding = 1\nn_components_for_compound_encoding = 25\nalpha_regularization_for_linear_models = 10\n\n# predict_method\n\nmodel_type = 'Ridge'","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:20:09.746990Z","iopub.execute_input":"2023-09-24T03:20:09.748030Z","iopub.status.idle":"2023-09-24T03:20:09.753749Z","shell.execute_reply.started":"2023-09-24T03:20:09.747991Z","shell.execute_reply":"2023-09-24T03:20:09.752346Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"str_model_id = model_type\nstr_model_id += ' nCT'+ str(n_components_for_cell_type_encoding)\nstr_model_id += ' nCD'+ str(n_components_for_compound_encoding)\nstr_model_id += ' Al'+ str(alpha_regularization_for_linear_models)\nstr_model_id += ' ' +str_inf_target_dimred+str( n_components )\n\nprint( str_model_id )","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:20:09.755612Z","iopub.execute_input":"2023-09-24T03:20:09.755995Z","iopub.status.idle":"2023-09-24T03:20:09.783583Z","shell.execute_reply.started":"2023-09-24T03:20:09.755963Z","shell.execute_reply":"2023-09-24T03:20:09.782341Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Target encoded features","metadata":{}},{"cell_type":"code","source":"%%time\n# Yr = reducer.fit_transform(X)\n# n_components_for_cell_type_encoding = 10\ndf_tmp = pd.DataFrame(Yr[:, :n_components_for_cell_type_encoding  ], index = df_de_train.index  )\ndf_tmp['column for aggregation'] = df_de_train['cell_type']\ndf_cell_type_encoded = df_tmp.groupby('column for aggregation').quantile( quantile )\nprint('df_cell_type_encoded.shape', df_cell_type_encoded.shape )\ndisplay( df_cell_type_encoded )\n\n\n# n_components_for_compound_encoding = 10\ndf_tmp = pd.DataFrame(Yr[:, :n_components_for_compound_encoding  ], index = df_de_train.index  )\ndf_tmp['column for aggregation'] = df_de_train['sm_name']\ndf_compound_encoded = df_tmp.groupby('column for aggregation').quantile( quantile )\nprint('df_compound_encoded.shape', df_compound_encoded.shape )\ndisplay( df_compound_encoded )","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:20:09.785814Z","iopub.execute_input":"2023-09-24T03:20:09.786181Z","iopub.status.idle":"2023-09-24T03:20:09.849566Z","shell.execute_reply.started":"2023-09-24T03:20:09.786151Z","shell.execute_reply":"2023-09-24T03:20:09.848339Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Prepare X_train, X_submit - target encoded cell type and compound features","metadata":{}},{"cell_type":"code","source":"%%time\nX_train = np.zeros( (len( df_de_train ) , n_components_for_cell_type_encoding + n_components_for_compound_encoding ))\n\nfor i in range(len( X_train )):\n    cell_type = df_de_train['cell_type'].iat[i] \n    X_train[i,:n_components_for_cell_type_encoding] = df_cell_type_encoded.loc[cell_type,:].values  \n    compound = df_de_train['sm_name'].iat[i] \n    X_train[i,n_components_for_cell_type_encoding:] = df_compound_encoded.loc[ compound, : ].values\nprint( X_train.shape)     \nprint( X_train)     \n    \n\nX_submit = np.zeros( (len( df_id_map ) , n_components_for_cell_type_encoding + n_components_for_compound_encoding ))\nfor i in range(len( X_submit )):\n    cell_type = df_id_map['cell_type'].iat[i] \n    X_submit[i,:n_components_for_cell_type_encoding] = df_cell_type_encoded.loc[cell_type,:].values  \n    compound = df_id_map['sm_name'].iat[i] \n    X_submit[i,n_components_for_cell_type_encoding:] = df_compound_encoded.loc[ compound, : ].values\n    \n    \nprint( X_submit.shape)     \nprint( X_submit)","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:20:09.851248Z","iopub.execute_input":"2023-09-24T03:20:09.851627Z","iopub.status.idle":"2023-09-24T03:20:10.070695Z","shell.execute_reply.started":"2023-09-24T03:20:09.851596Z","shell.execute_reply":"2023-09-24T03:20:10.069607Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Modeling","metadata":{}},{"cell_type":"code","source":"%%time\nfrom sklearn.linear_model import Ridge\n\nmodel = Ridge(alpha=alpha_regularization_for_linear_models)\nprint(model)\nmodel.fit(X_train, Yr)\n\nY_submit = reducer.inverse_transform(   model.predict(X_submit) )\nprint(Y_submit.shape)\nY_submit","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:20:10.072781Z","iopub.execute_input":"2023-09-24T03:20:10.073229Z","iopub.status.idle":"2023-09-24T03:20:10.124122Z","shell.execute_reply.started":"2023-09-24T03:20:10.073189Z","shell.execute_reply":"2023-09-24T03:20:10.122453Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"# Save submission CSV","metadata":{}},{"cell_type":"code","source":"%%time\ndf_submit = pd.DataFrame(Y_submit, columns = df_de_train.columns[5:])\ndf_submit.index.name = 'id'\nprint( df_submit.shape )\ndisplay(df_submit)\ndf_submit.to_csv('submission.csv')","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:20:10.126558Z","iopub.execute_input":"2023-09-24T03:20:10.127313Z","iopub.status.idle":"2023-09-24T03:20:23.109673Z","shell.execute_reply.started":"2023-09-24T03:20:10.127245Z","shell.execute_reply":"2023-09-24T03:20:23.108504Z"},"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"print('%.1f seconds passed total '%(time.time()-t0start) )\nprint('%.1f minutes passed total '%( (time.time()-t0start)/60)  )\nprint('%.2f hours passed total '%( (time.time()-t0start)/3600)  )","metadata":{"execution":{"iopub.status.busy":"2023-09-24T03:20:23.111637Z","iopub.execute_input":"2023-09-24T03:20:23.111985Z","iopub.status.idle":"2023-09-24T03:20:23.118228Z","shell.execute_reply.started":"2023-09-24T03:20:23.111955Z","shell.execute_reply":"2023-09-24T03:20:23.117086Z"},"trusted":true},"execution_count":null,"outputs":[]}]}